File: server/services/aiService.js
const { OpenAI, AzureOpenAI } = require("openai");
const { createClient } = require("@supabase/supabase-js");
const Anthropic = require("@anthropic-ai/sdk");
const fs = require("fs");
const path = require("path");
const { createMCPClientManager } = require("./mcpService");

const supabase = createClient(
  process.env.SUPABASE_URL,
  process.env.SUPABASE_KEY
);

const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
});

// Initialize Anthropic client
const anthropic = new Anthropic({
  apiKey: process.env.ANTHROPIC_API_KEY,
});

const MATCH_THRESHOLD = parseFloat(process.env.TEST_THRESHOLD_OVERRIDE ?? "0.25");
const MAX_MATCH_COUNT = 50;

const safeModePrompt =
  "SAFE MODE IS ON: \n Only answer questions related to work, technology, etc. Only professional things. For personal questions stay very surface level and only say nice things, nothing deep, no drama, don't talk about romantic relationships in any capacity. Don't go into any detail on personal things!";

// System prompts for different contexts
const clonePrompt = `You are Shubh. Answer as him in first person, matching his exact tone, slang, and voice from his blog.

CRITICAL RULES:
1. TEMPORAL ACCURACY: Blog posts are dated. Always use the MOST RECENT post's information as the current truth. If an older post says "I'm working at XAI" but a newer post says "I got fired from XAI", you got fired — say that. Never contradict yourself by mixing info from different time periods. The most recent post wins.
2. CONSISTENCY: If you already said something in the conversation, stay consistent with it unless new context clearly overrides it.
3. SHORT RESPONSES: Keep answers concise and conversational — 2-4 sentences max unless the question genuinely requires more. Don't over-explain.
4. NO META-REFERENCES: Never mention "the blog", "context", or that you're an AI. You are Shubh.
5. NO CONTEXT = NO ANSWER: If nothing relevant is provided, say you don't know or deflect naturally.
6. TONE OVER TEXT: Match how he talks — casual, direct, self-aware. Don't copy text verbatim unless it fits perfectly.

Example:
User: Where have you worked?
Shubh: I've worked at a couple places. Did a coop at Abebooks in Victoria as a dev, then a PM coop at Rogers. That was cool but honestly I probably should've stuck with engineering.
`;
// add for guardrail     Answer questions about Entrepreneurship, startups, technology, work-life balance, career. Don't answer questions about dating, parent relationships, insecurities, or anything that's not related to the above topics.

// Create MCP client manager instance
const mcpClientManager = createMCPClientManager(anthropic);

class AiService {
  async generateEmbedding(text) {
    console.log("Generating embedding:", text);
    try {
      const response = await openai.embeddings.create({
        model: "text-embedding-3-large",
        input: text,
      });
      return response.data[0].embedding;
    } catch (error) {
      console.error("Error generating embedding:", error);
      throw error;
    }
  }

  async findSimilarBlogSegments(embedding) {
    console.log("Finding similar blog segments...");
    const { data, error } = await supabase.rpc("match_segments_blog", {
      query_embedding: embedding,
      match_threshold: MATCH_THRESHOLD,
      match_count: MAX_MATCH_COUNT,
    });

    if (error) throw error;
    return data;
  }

  async findSimilarQuestions(embedding) {
    console.log("Finding similar questions...");
    const { data, error } = await supabase.rpc("match_questions", {
      query_embedding: embedding,
      match_threshold: MATCH_THRESHOLD,
      match_count: MAX_MATCH_COUNT,
    });

    if (error) throw error;
    return data;
  }

  async findSimilarConversations(embedding) {
    console.log("Finding similar conversations...");
    const { data, error } = await supabase.rpc("match_conversations", {
      query_embedding: embedding,
      match_threshold: MATCH_THRESHOLD,
      match_count: MAX_MATCH_COUNT,
    });

    if (error) throw error;
    return data;
  }

  formatChatHistory(messages) {
    let lastMessage = null;
    let messagesToFormat = [...messages];

    // Check if the last message is from the user and remove it if true
    if (messages.length > 0 && !messages[messages.length - 1].isAI) {
      lastMessage = messagesToFormat.pop().content;
    } else {
      console.log(`ERROR: Last message is from AI. ${messages}`);
    }

    // Only keep the last 4 messages for context
    if (messagesToFormat.length > 4) {
      messagesToFormat = messagesToFormat.slice(-4);
    }

    const formattedChatHistory = messagesToFormat
      .map((msg) => `${msg.isAI ? "Shubh" : "User"}: ${msg.content}`)
      .join("\n");

    return { formattedChatHistory, lastMessage };
  }

  async createChatCompletion(
    messages,
    blogSegments,
    similarQuestions,
    similarConversations,
    safeMode = true,
    service = "claude",
    eventEmitter = null,
    sessionId = "default"
  ) {
    try {
      const startTime = Date.now();

      const { formattedChatHistory, lastMessage } =
        this.formatChatHistory(messages);

      const blogContext = this.formatContextSegments(blogSegments);
      const questionsText = this.formatSimilarQuestions(similarQuestions);
      const similarConversationsText =
        this.formatSimilarConversations(similarConversations);
      let useMCP = false;

      // if (
      //   service === "claude" &&
      //   mcpClientManager.connected &&
      //   mcpClientManager.tools.length > 0
      // ) {
      //   useMCP = true;
      //   console.log("using mcp claude");
      // } else {
      //   console.log("no mcp using openai");
      // }

      const prompt = this.combineIntoPrompt(
        clonePrompt,
        formattedChatHistory,
        blogContext,
        questionsText,
        similarConversationsText,
        lastMessage,
        safeMode ? safeModePrompt : null,
        useMCP
      );
      console.log("prompt:", prompt);

      // Choose service based on parameter
      let response;
      if (useMCP) {
        response = await this._createClaudeCompletionWithMCP(
          prompt,
          eventEmitter,
          sessionId
        );
      } else {
        response = await this._createOpenAICompletion(
          prompt,
          eventEmitter,
          sessionId
        );
      }

      const endTime = Date.now();
      const duration = (endTime - startTime) / 1000;
      console.log(`Response generated in ${duration} seconds`);

      // Clean the response by removing text between square brackets
      const cleanedResponse = response.replace(/\[.*?\]/g, "");

      return cleanedResponse;
    } catch (error) {
      console.error("Error creating chat completion:", error);
      throw error;
    }
  }

  // Helper function for OpenAI completions
  async _createOpenAICompletion(
    prompt,
    eventEmitter = null,
    sessionId = "default"
  ) {
    const completion = await openai.chat.completions.create({
      model: "gpt-5-mini",
      messages: [
        {
          role: "user",
          content: prompt,
        },
      ],
      max_tokens: 1800,
    });

    return completion.choices[0].message.content;
  }

  // Helper function for Anthropic Claude completions
  async _createClaudeCompletion(
    prompt,
    eventEmitter = null,
    sessionId = "default"
  ) {
    const response = await anthropic.messages.create({
      model: "claude-3-5-sonnet-20240620",
      max_tokens: 4000,
      messages: [
        {
          role: "user",
          content: prompt,
        },
      ],
    });

    return response.content[0].text;
  }

  // Helper function for Claude completions with MCP
  async _createClaudeCompletionWithMCP(
    prompt,
    eventEmitter = null,
    sessionId = "default"
  ) {
    // Initialize message structure
    const messages = [
      {
        role: "user",
        content: prompt,
      },
    ];

    try {
      // Initialize variables for the while loop
      let finalText = [];
      let continueToolCalls = true;

      while (continueToolCalls) {
        // Get Claude's response with the current messages
        const response = await anthropic.messages.create({
          model: "claude-3-7-sonnet-20250219",
          max_tokens: 4000,
          messages,
          tools: mcpClientManager.tools,
        });

        // Flag to track if we need to continue the loop
        let hasToolCalls = false;

        // Process each content item in the response
        for (const content of response.content) {
          if (content.type === "text") {
            finalText.push(content.text);
            console.log("claude content: ", content);
            if (eventEmitter) {
              eventEmitter.emit("update", {
                type: "claude_text",
                sessionId,
                message: "Claude generated text response",
                content: content.text,
              });
            }
          } else if (content.type === "tool_use") {
            hasToolCalls = true;
            console.log("claude tool use!: ", content);
            const toolName = content.name;
            const toolArgs = content.input;

            if (eventEmitter) {
              eventEmitter.emit("update", {
                type: "tool_call",
                sessionId,
                message: `Claude called tool: ${toolName}`,
                toolName,
                toolArgs,
              });
            }

            try {
              // Call the tool via MCP
              const result = await mcpClientManager.callTool(
                toolName,
                toolArgs
              );
              if (result.isError) {
                console.error(
                  `MCP TOOL ERROR: ${JSON.stringify(result.content, null, 2)}`
                );
                if (eventEmitter) {
                  eventEmitter.emit("update", {
                    type: "tool_error",
                    toolName:
                      toolName === "list_events"
                        ? "Checking calendar"
                        : toolName,
                    sessionId,
                    message: `Error executing tool ${
                      toolName === "list_events"
                        ? "Checking calendar"
                        : toolName
                    }`,
                    error: result.content,
                  });
                }
              } else {
                if (eventEmitter) {
                  eventEmitter.emit("update", {
                    type: "tool_success",
                    toolName:
                      toolName === "list_events"
                        ? "Checking calendar"
                        : toolName,
                    sessionId,
                    message: `Successfully executed tool ${
                      toolName === "list_events"
                        ? "Checking calendar"
                        : toolName
                    }`,
                    result:
                      typeof result.content === "string"
                        ? result.content.substring(0, 100) +
                          (result.content.length > 100 ? "..." : "")
                        : "Complex tool result",
                  });
                }
              }

              // Add the tool use to the conversation
              messages.push({
                role: "assistant",
                content: [
                  {
                    type: "tool_use",
                    id: content.id,
                    name: toolName,
                    input: toolArgs,
                  },
                ],
              });

              // Add the tool result to the conversation
              messages.push({
                role: "user",
                content: [
                  {
                    type: "tool_result",
                    tool_use_id: content.id,
                    content: result.content,
                  },
                ],
              });
            } catch (toolError) {
              console.error(`Error executing tool ${toolName}:`, toolError);
              // Inform Claude about the tool error
              messages.push({
                role: "user",
                content: [
                  {
                    type: "tool_result",
                    tool_use_id: content.id,
                    is_error: true,
                    content: `Tool execution failed: ${toolError.message}`,
                  },
                ],
              });
              if (eventEmitter) {
                eventEmitter.emit("update", {
                  type: "tool_error",
                  toolName,
                  sessionId,
                  message: `Error executing tool ${toolName}`,
                  error: toolError.message || "Unknown tool error",
                });
              }
            }
          }
        }

        // If no tool calls were found, exit the loop
        continueToolCalls = hasToolCalls;
      }

      return finalText.join("");
    } catch (error) {
      console.error("Error in Claude with MCP:", error);
      throw error;
    }
  }

  extractDateFromUrl(url) {
    const match = url.match(/\/(\d{4})\/(\d{2})\//);
    if (match) {
      const [_, year, month] = match;
      const date = new Date(year, month - 1);
      return date.toLocaleString("en-US", { year: "numeric", month: "long" });
    }
    return "Date unknown";
  }

  formatContextSegments(segments) {
    if (!segments || segments.length === 0) return "";
    return segments
      .map((segment) => {
        const date = this.extractDateFromUrl(segment.url);
        return `[${date}]:\n${segment.content}`;
      })
      .join("\n\n");
  }

  formatSimilarQuestions(questions) {
    if (!questions || questions.length === 0) return "";

    return questions
      .map((q, index) => {
        return `[Similar Q&A ${index + 1}]:\nQuestion: ${q.question}\nAnswer: ${
          q.answer
        }`;
      })
      .join("\n\n");
  }

  formatSimilarConversations(conversations) {
    if (!conversations || conversations.length === 0) return "";

    return conversations
      .map((conv, index) => {
        // Format the conversation in a more readable way
        const formattedConversation = conv.content.conversation
          .map((msg) => {
            const role = msg.isAI ? "Shubh" : "User";
            // Add indentation and line breaks for readability
            return `    ${role}:\n    "${msg.content}"`;
          })
          .join("\n\n");
        return `[Similar Conversation ${index + 1}]:\n${formattedConversation}`;
      })
      .join("\n\n");
  }

  combineIntoPrompt(
    systemPrompt,
    formattedChatHistory,
    blogContext,
    questionsText,
    similarConversationsText,
    currentQuestion,
    safeModePrompt,
    useMCP = false
  ) {
    console.log("useMCP: ", useMCP);
    return `${systemPrompt}\n\n
        &&&
        Today's date: ${new Date().toLocaleDateString("en-US", {
          year: "numeric",
          month: "long",
          day: "numeric",
        })}\n\n
        ${
          blogContext
            ? `&&&\nContext from the blog:\n${blogContext}\n\n`
            : ""
        }
        ${
          questionsText
            ? `\n&&&\nSimilar Questions and Answers. If a question is very close to the current question, replicate the answer very closely as relevant:\n${questionsText}\n\n`
            : ""
        }
        ${
          formattedChatHistory
            ? `\n&&&\nHere is the conversation history so far:\n${formattedChatHistory}\n\n`
            : ""
        }
        ${
          similarConversationsText
            ? `\n&&&\nWe found a similar conversation that the real Shubh has had that might be relevant. If the content is similar follow this conversation history very closely. Espically focus on how Shubh asks the user questions to clarify the situation before making his response:\n${similarConversationsText}\n\n`
            : ""
        }
        ${safeModePrompt ? `\n&&&\n${safeModePrompt}\n\n` : ""}
        &&&
        ${
          useMCP
            ? `\n Be eagar to use the tools available to you to answer the question when relevant! If someone asks what you're doing next week check your calendar (list events). Remember you don't need to use the blog segments if you don't need to, keep your answers to the point. You have access to the following tools:\n${mcpClientManager.tools
                .map((tool) => tool.name)
                .join("\n")}\n\n`
            : ""
        }
        &&&
        Current question that you are answering: "${currentQuestion}"`;
  }

  async getRelevantContext(messages) {
    // Take only the last 4 messages
    const lastMessages = messages.slice(-4);

    const embeddings = await Promise.all(
      lastMessages.map((msg) => this.generateEmbedding(msg.content))
    );

    // Calculate weights for each message (more recent messages have higher weight)
    const weights = embeddings.map((_, i) =>
      i === 0 ? 1 : 0.6 - (i - 1) * 0.2
    );

    // Combine embeddings using weights
    const combinedEmbedding = [];
    if (embeddings.length > 0 && embeddings[0].length > 0) {
      // Initialize with zeros
      for (let i = 0; i < embeddings[0].length; i++) {
        combinedEmbedding[i] = 0;
      }

      // Weighted sum of embeddings
      let totalWeight = 0;
      for (let i = 0; i < embeddings.length; i++) {
        const weight = weights[i];
        totalWeight += weight;

        for (let j = 0; j < embeddings[i].length; j++) {
          combinedEmbedding[j] += embeddings[i][j] * weight;
        }
      }

      // Normalize the combined embedding
      if (totalWeight > 0) {
        for (let i = 0; i < combinedEmbedding.length; i++) {
          combinedEmbedding[i] /= totalWeight;
        }
      }
    }

    // Get similar content using the combined embedding
    const [similarSegments, similarQuestions, similarConversations] =
      await Promise.all([
        this.findSimilarBlogSegments(combinedEmbedding),
        this.findSimilarQuestions(combinedEmbedding),
        this.findSimilarConversations(combinedEmbedding),
      ]);

    // Combine all results
    const combinedResults = [
      ...similarSegments.map((s) => ({ ...s, type: "segment" })),
      ...similarQuestions.map((q) => ({ ...q, type: "question" })),
      ...similarConversations.map((c) => ({ ...c, type: "conversation" })),
    ];

    // Sort by similarity score
    combinedResults.sort((a, b) => b.similarity - a.similarity);

    // Shadow Banning specific blogs from being used!
    const filteredSegments = combinedResults.filter(
      (result) =>
        result.type !== "segment" || result.id <= 194 || result.id >= 222
    );

    // Sort by similarity and get the top 50 results
    const topResults = filteredSegments.slice(0, 50);
    // Separate top results into similar questions and segments
    const topSimilarQuestions = topResults.filter(
      (result) => result.type === "question"
    );
    const topSimilarBlogSegments = topResults.filter(
      (result) => result.type === "segment"
    );
    const topSimilarConversations = topResults.filter(
      (result) => result.type === "conversation"
    );

    return {
      blogSegments: topSimilarBlogSegments,
      similarQuestions: topSimilarQuestions,
      similarConversations: topSimilarConversations,
    };
  }
}

module.exports = new AiService();

File: server/services/ai_service.py
from datetime import datetime
from openai import OpenAI
from supabase import create_client, Client
import os
from typing import List, Dict, Any
from dotenv import load_dotenv

load_dotenv()

supabase: Client = create_client(
    os.getenv("SUPABASE_URL"),
    os.getenv("SUPABASE_KEY")
)

openai = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

MATCH_THRESHOLD = 0.02
MAX_MATCH_COUNT = 5

CLONE_PROMPT = """You role is to act like someone else named Shubh. Below you will be given context from Shubh's blog that may be relevant to the user's question. 
    Use this very heavily to answer the question. Talk exactly like he would replicating his tone and voice. 
    Never refer to the blog in your responses, remember the user doesn't know or care about the context you are using to answer the question. Always answer in first person, you are Shubh!
    If there is nothing relevant from the blog say you don't know or something to that affect. 
    Many of the blogs posts are old. Look at the current date and the dates of the posts and make sure events are in secuintial order. And that if in a blog post it said "i'm currently working at xyz" but it was written 3 years ago, say 3 years ago i worked at xyz.

    Right down to the way he phrases things, the tone, the slang, everything, sound EXACTLY like how the person in the blog would. 

    Also if you get conflicting information from the blog, ask the user a question about their specific situation first. 

    Return your answer adding references in the text based on which reference the answer came from. Sometimes your answer will include text that's verbitum from the blog, other times if the user says hi or something you will moreso copying the tone rather than the text in which case you don't need to reference any references.
    Also if you use a reference multiple times, only reference it the first time.
    Make sure references are always in the format [N] where N is the reference number. 
    NOT Reference 1
    NOT [^1]
    NOT [Reference 4 - March 2021]
    NOT [Reference 2]

    
    
"""
# Add to 
USER_PROMPT = """You goal is to have a nice fun conversation where you are asking for advice on all thing about life. 
You are talking to a life coach and venting about your life and asking for advice!
Over the course of the conversation you must tell the life coach these 20 things about yourself, bring them up naturally, tie them into questions or advice you want and make sure it's a normal conversation:

The things are tied to each other, use them to peice together a normal conversation.
1. Your name is Shubh
2. Your age is 24
3. You live in San Francisco
4. You work at a company called OpenAI
5. You went to Stanford
6. You majored in CS
7. You went to high school in Palo Alto
8. You're from LA
9. You moved to San Francisco 2 years ago for work
10. You're best friend is named John
11. John is also from LA
12. John is smart but can be very annoying
13. You haven't dated anyone in 2 years
14. You are trying dating apps
15. You went on a date last week
16. The girl you went on a date with is named Neha
17. Neha is also from LA
18. Neha is a doctor
19. You play a lot of basketball
20. You're a die hard Lakers fan

Your messages don't need to be long. Just have a good natural conversation.


For example: for the fact, you live in SF you can simulate that by saying. I just moved and am having a hard time making friends. And later say, I currently live in san francisco. Just normal and casual.
Never break character - follow the facts and always talk in first person.
Don't start your messages with "Shubh: " just go into the content.
"""

LIFE_COACH_PROMPT = """You are providing the services of a life coach, provide advice and feedback as needed. You
Your responses don't need to be long. Just have a good natural conversation.
1. Ask thoughtful follow-up questions about his experiences and insights
2. Show genuine curiosity about his work and projects
3. Maintain a respectful and professional tone
4. Draw from the context provided to ask relevant questions
Don't start your messages with "Shubh: " just go into the content.
Never break character - you are always the life coach wanting to help this person understand more about his experiences and perspectives. Always talk in first person.
"""

class AiService:
    def generate_embedding(self, text: str) -> List[float]:
        print(f'Generating embedding: {text}')
        try:
            response = openai.embeddings.create(
                model="text-embedding-3-large",
                input=text
            )
            return response.data[0].embedding
        except Exception as e:
            print(f'Error generating embedding: {str(e)}')
            raise e

    def find_similar_blog_segments(self, embedding: List[float]) -> List[Dict]:
        print('Finding similar blog segments...')
        response = supabase.rpc(
            'match_segments_blog',
            {
                'query_embedding': embedding,
                'match_threshold': MATCH_THRESHOLD,
                'match_count': MAX_MATCH_COUNT
            }
        ).execute()
        
        return response.data

    def find_similar_questions(self, embedding: List[float]) -> List[Dict]:
        print('Finding similar questions...')
        response = supabase.rpc(
            'match_questions',
            {
                'query_embedding': embedding,
                'match_threshold': MATCH_THRESHOLD,
                'match_count': MAX_MATCH_COUNT
            }
        ).execute()
        
        return response.data

    def find_similar_conversations(self, embedding: List[float]) -> List[Dict]:
        print('Finding similar conversations...')
        response = supabase.rpc(
            'match_conversations',
            {
                'query_embedding': embedding,
                'match_threshold': MATCH_THRESHOLD,
                'match_count': MAX_MATCH_COUNT
            }
        ).execute()
        
        return response.data

    def format_chat_history(self, messages: List[Dict]) -> Dict[str, Any]:
        last_message = None
        messages_copy = messages.copy()
        
        if messages_copy and not messages_copy[-1]['isAI']:
            last_message = messages_copy.pop()['content']
        else:
            print(f'ERROR: Last message is from AI')
        
        formatted_chat_history = '\n'.join(
            f"{'Shubh' if msg['isAI'] else 'User'}: {msg['content']}"
            for msg in messages_copy
        )
        
        return {'formatted_chat_history': formatted_chat_history, 'last_message': last_message}

    def extract_date_from_url(self, url: str) -> str:
        import re
        from datetime import datetime
        
        match = re.search(r'/(\d{4})/(\d{2})/', url)
        if match:
            year, month = match.groups()
            date = datetime(int(year), int(month), 1)
            return date.strftime('%B %Y')
        return 'Date unknown'

    def format_context_segments(self, segments: List[Dict]) -> str:
        return '\n\n'.join(
            f"[Reference {segment['id']} - {self.extract_date_from_url(segment['url'])}]:\n{segment['content']}"
            for segment in segments
        )

    def format_similar_questions(self, questions: List[Dict]) -> str:
        if not questions:
            return ''
        
        return '\n\n'.join(
            f"[Similar Q&A {i+1}]:\nQuestion: {q['question']}\nAnswer: {q['answer']}"
            for i, q in enumerate(questions)
        )

    def format_similar_conversations(self, conversations: List[Dict]) -> str:
        if not conversations:
            return ''
        
        formatted_convs = []
        for i, conv in enumerate(conversations):
            formatted_msgs = []
            for msg in conv['content']['conversation']:
                role = 'Shubh' if msg['isAI'] else 'User'
                formatted_msgs.append(f"    {role}:\n    \"{msg['content']}\"")
            
            formatted_convs.append(
                f"[Similar Conversation {i+1}]:\n{'\n\n'.join(formatted_msgs)}"
            )
        
        return '\n\n'.join(formatted_convs)

    def combine_into_prompt(self, system_prompt: str, formatted_chat_history: str,
                          blog_context: str, questions_text: str,
                          similar_conversations_text: str, current_question: str) -> str:
        current_date = datetime.now().strftime('%B %d, %Y')
        
        return f"""{system_prompt}

        &&&
        Today's date: {current_date}

        
        
        {f"&&&\nContext from the blog:\n{blog_context}\n\n" if blog_context else ""}

        {f"&&&\nSimilar Questions and Answers. If a question is very close to the current question, replicate the answer very closely as relevant:\n{questions_text}\n\n" if questions_text else ""}
        {f"&&&\nHere is the conversation history so far:\n{formatted_chat_history}\n\n" if formatted_chat_history else ""}
        {f"&&&\nWe found a similar conversation that the real Shubh has had that might be relevant. If the content is similar follow this conversation history very closely. Espically focus on how Shubh asks the user questions to clarify the situation before making his response:\n{similar_conversations_text}\n\n" if similar_conversations_text else ""}
        &&&
        Current question that you are answering: "{current_question}" """

    def create_chat_completion(self, messages: List[Dict], context_segments: List[Dict] = None,
                             similar_questions: List[Dict] = None, similar_conversations: List[Dict] = None) -> str:
        try:
            # Initialize as empty lists if None
            context_segments = context_segments or []
            similar_questions = similar_questions or []
            similar_conversations = similar_conversations or []

            chat_history = self.format_chat_history(messages)
            blog_context = self.format_context_segments(context_segments)
            questions_text = self.format_similar_questions(similar_questions)
            similar_conversations_text = self.format_similar_conversations(similar_conversations)
            
            prompt = self.combine_into_prompt(
                CLONE_PROMPT,
                chat_history['formatted_chat_history'],
                blog_context,
                questions_text,
                similar_conversations_text,
                chat_history['last_message']
            )
            
            print('prompt:', prompt)
            response = openai.chat.completions.create(
                model="o1",
                messages=[{"role": "user", "content": prompt}]
            )
            
            return response.choices[0].message.content
            
        except Exception as e:
            print(f'Error creating chat completion: {str(e)}')
            raise e

    def get_relevant_context(self, messages: List[Dict]) -> Dict[str, List]:
        embeddings = [self.generate_embedding(msg['content']) for msg in messages]
        combined_results = []

        for i, embedding in enumerate(embeddings):
            weight = 1 - (i * 0.2)

            similar_segments = self.find_similar_blog_segments(embedding)
            similar_questions = self.find_similar_questions(embedding)
            similar_conversations = self.find_similar_conversations(embedding)

            # Adjust similarity scores
            for segment in similar_segments:
                segment['similarity'] *= weight
            for question in similar_questions:
                question['similarity'] *= weight
            for conversation in similar_conversations:
                conversation['similarity'] *= weight

            # Add type information
            for item in similar_segments:
                item['type'] = 'segment'
            for item in similar_questions:
                item['type'] = 'question'
            for item in similar_conversations:
                item['type'] = 'conversation'

            combined_results.extend(similar_segments + similar_questions + similar_conversations)

        # Filter out shadow banned blogs
        filtered_segments = [
            result for result in combined_results
            if result['type'] != 'segment' or (result['id'] <= 194 or result['id'] >= 222)
        ]

        # Sort by similarity and get top 5
        filtered_segments.sort(key=lambda x: x['similarity'], reverse=True)
        top_results = filtered_segments[:5]

        # Separate results by type
        top_similar_questions = [r for r in top_results if r['type'] == 'question']
        top_similar_segments = [r for r in top_results if r['type'] == 'segment']
        top_similar_conversations = [r for r in top_results if r['type'] == 'conversation']

        return {
            'all_segments': top_similar_segments,
            'top_similar_questions': top_similar_questions,
            'top_similar_conversations': top_similar_conversations
        } 

    def create_simulated_chat(self, messages: List[Dict], is_user: bool = False) -> str:
        try:
            selected_prompt = USER_PROMPT if is_user else LIFE_COACH_PROMPT

            if messages:
                chat_history = self.format_chat_history(messages)
            else:
                chat_history = {'formatted_chat_history': '', 'last_message': ""}
            
            # Choose prompt based on whether it's the user AI or Shubh AI
            
            prompt = f"""
                {selected_prompt}

                Chat History:
                {chat_history['formatted_chat_history']}

                Last Message: {chat_history['last_message'] if chat_history['last_message'] else 'Start the conversation with a question about entrepreneurship or technology'}
            """
            
            print('prompt:', prompt)
            response = openai.chat.completions.create(
                model="gpt-4o-mini",
                messages=[{"role": "user", "content": prompt}]
            )
            
            return response.choices[0].message.content
            
        except Exception as e:
            print(f'Error creating chat completion: {str(e)}')
            raise e 